astropy/io/ascii/tests/test_tdat.py

```diff
@@ -0,0 +1,394 @@
+    "field[id] = int4  [meta.id] (index) // Source ID Number",
+    "field[name] = char12  [meta.id;meta.main] (index) // String Name",
+    "field[ra] = float8:.4f_degree [pos.eq.ra] (index) // Right Ascension",
+    "field[dec] = float8:.4f_degree [pos.eq.dec] (index) // Declination",
+    "field[empty] = float8:.4f // Empty // Comment",
+    "#",
+    "parameter_defaults = name ra dec",
+    "#",
+    "# Virtual Parameters",
+    "#",
+    "frequency_regime = Gamma-ray",
+    "observatory_name = GAMMA-RAY BURSTS",
+    "row_type = GRB",
+    "table_author = Example et al.",
+    "table_priority = 3.01",
+    "table_type = Observation",
+    "unique_key = record_number",
+    "#",
+    "# Data Format Specification",
+    "#",
+    "line[1] = record_number id name ra dec empty",
+    "#",
+    "<DATA>",
+    "1|10|aaa|1.0|1.0||",
+    "2|20|b|2.0|||",
+    "3|30|c||3.0||",
+    "4|20|||||",
+    "5||||||",
+    "|60|f|6.0|6.0||",
+    "7| 70| g | 7.0 |7.0||",
+    "<END>",
+]
+test_table = Table.read(test_dat, format="ascii.tdat")
+# Corresponds to simple_table()
+SIMPLE_LINES = [
+    "<HEADER>",
+    "table_name = astropy_table",
+    "#",
+    "# Table Parameters",
+    "#",
+    "field[a] = int4",
+    "field[b] = float8",
+    "field[c] = char1",
+    "#",
+    "# Data Format Specification",
+    "#",
+    "line[1] = a b c",
+    "#",
+    "<DATA>",
+    "1|1.0|c|",
+    "2|2.0|d|",
+    "3|3.0|e|",
+    "<END>",
+]
+
+
+def test_catch_format():
+    """
+    Ensure that a table without a table_name specified issues a warning.
+    """
+    t = simple_table()
+    out = StringIO()
+    with pytest.warns(TdatFormatWarning, match="'table_name' must be specified"):
+        t.write(out, format="ascii.tdat")
+
+    t.meta["table_name"] = "anastropytablewithanamethatistoolongfortheformat"
+    with pytest.warns(TdatFormatWarning, match="'table_name' is too long"):
+        t.write(out, format="ascii.tdat")
+
+    t.meta["table_name"] = "astropy_table"
+    t.meta["table_description"] = """
+    This is a description that exceeds the character limit allowed by the tdat
+    format and it should be truncated before being written. A warning should
+    pop up to inform the user of this behavior.
+    """
+    with pytest.warns(TdatFormatWarning, match="'table_description' is too long"):
+        t.write(out, format="ascii.tdat")
+
+
+def test_write_simple():
+    """
+    Write a simple table with common types.  This shows the compact version
+    of serialization with one line per column.
+    """
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", category=TdatFormatWarning)
+        t = simple_table()
+        out = StringIO()
+        t.write(out, format="ascii.tdat")
+        assert out.getvalue().splitlines() == SIMPLE_LINES
+
+
+def test_write_full():
+    """
+    Write a full-featured table with common types and explicitly check output
+    """
+    t = test_table
+    lines = [
```

Comment:
Is `lines` here exactly the same as `test_dat`? If not, what are the differences and why?

Reply:
There are a couple differences
* Empty comment lines are dropped (except when denoting a section header like "Table Parameters"). 
* The data type for string field "name" is downsized to char3 from char12 reflecting the maximum size in the column. 
* Extraneous spaces in the data are stripped